src/app/checklist/app-updates/app-updates.service.ts

```diff
@@ -0,0 +1,51 @@
+import { Injectable } from '@angular/core';
+import { MatSnackBar } from '@angular/material';
+import { SwUpdate } from '@angular/service-worker';
+import {
+  fromEvent as observableFromEvent,
+  interval,
+  merge as observableMerge,
+  Observable,
+  of as observableOf
+} from 'rxjs';
+import { map, switchMap, tap } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AppUpdatesService {
+  constructor(public snackbar: MatSnackBar, private serviceWorkerUpdates: SwUpdate) {}
+
+  checkUpdatesAuto() {
+    this.serviceWorkerUpdates.activated.subscribe(() => {
+      this.snackbar.open('App updated', 'OK', { duration: 2000 });
+    });
+
+    this.serviceWorkerUpdates.available
+      .pipe(
+        switchMap(() => {
+          const updateSnackbar = this.snackbar.open('Update available!', 'Update Now');
+          return updateSnackbar.onAction().pipe(tap(() => updateSnackbar.dismiss()));
+        })
+      )
+      .subscribe(() => window.location.reload());
+  }
+
+  checkUpdatesManually() {
+    this.snackbar.open(`You'll be notified if an update exists...`, 'OK', {
+      duration: 4000
+    });
+
+    interval(6 * 60 * 60).subscribe(() => {
+      this.serviceWorkerUpdates.checkForUpdate();
+    });
+  }
+
+  checkConnectionStatus(): Observable<boolean> {
+    return observableMerge(
+      observableOf(navigator.onLine),
+      observableFromEvent(window, 'online').pipe(map(() => true)),
```

Comment:
A tiny detail but you can use the `mapTo` operator if you are going to map to a fixed value.

Reply:
Thanks for the suggestions.

Yeah, so we have two options for checking for updates. The **automatic** check which will run in the app.component.

The **manual** update check, which I assume will be triggered by a button click exists as a safety net in case the service worker runs into update issues (it was expressed in the issue that the angular service worker isn't always 100% reliable). That way a user can escape being stuck on an older version by using that button to force the service worker into running the update check.

Reply:
Alright this is what I have in mind for the manual update check now:

1. When the method is called a dialog with the title **"checking for updates**" will popup and also show an indeterminate spinner.

2. Once the update check completes the spinner will hide and in it's place there will be text saying "The update check has completed, if a new version exists you will be notified soon"

3. The user then clicks "OK" to dismiss the dialog

Reply:
To prevent multiple subscriptions every time a user clicks the button:

- Keep a reference to the **checkForUpdates** Observable

- Every time the user clicks the button, Check if the reference exists and if it does exist unsubscribe to the old instance before subscribing to a new one.

Reply:
That sounds reasonable to me! A triny remark: You don't have to unsubscribe from the existing one and create a new interval if one exists already. So basically the logic would be as follows:
```
checkForUpdates() {
    if (!this.update$) {
      // Set up update interval
      this.update$ = <your-update-stream-here>.pipe(
        shareReplay(CACHE_SIZE) // make sure to use 'shareReplay' to multicast the interval
      );
    }

    return this.update$;
  }
```

Reply:
> That sounds reasonable to me! A triny remark: You don't have to unsubscribe from the existing one and create a new interval if one exists already. So basically the logic would be as follows:
> 
> ```
> checkForUpdates() {
>   if (!this.update$) {
>     // Set up update interval
>     this.update$ = <your-update-stream-here>.pipe(
>       shareReplay(CACHE_SIZE) // make sure to use 'shareReplay' to multicast the interval
>     );
>   }
> 
>   return this.update$;
> }
> ```
> Does that make sense?

Hmm I have to look further into this one to understand it, before I do though. I realised the the **checkForUpdate** method for the angular service-worker returns a **promise** and not an observable, the promise indicates whether the check has been completed or not. 

So I removed the observable interval and worked with just the promise, what do you think?

Reply:
If you expose this to the user then please convert this into an Observable using `fromPromise`. I think it would be reasonable to remove the interval and perform the check only once if that's the case already. I think I like it. I think this should be enough as a safety net. What do you think @KwintenP?

Reply:
@d3lm
Awesome, I'm not sure why we need to convert it back into an observable though?

here is the current implementation, maybe you can show me how you would adjust it?

```
  checkUpdatesManually() {
    const updateCheckSnackbar = this.snackbar.open(`Checking for updates...`);

    this.serviceWorkerUpdates
      .checkForUpdate()
      .then(() => {
        updateCheckSnackbar.dismiss();

        this.snackbar.open(`Check complete, You'll be notified if an update exists...`, 'OK', { duration: 2000 });
      })
      .catch(() => {
        updateCheckSnackbar.dismiss();

        this.snackbar.open(`Check could not complete, try again later`, 'OK', { duration: 2000 });
      });
  }
```

Reply:
While this works I'd like to avoid mixing and matching Promises and Observables. Observables are IMHO superior to Promises and it would be confusing to have them both. I really just wanna stick to one paradigm here and we are using Observables in many places throughout the entire application. So just pluck in a simple `fromPromise` and we are good.

Reply:
BTW, did you see @KwintenP's feedback? It seems you haven't really implemented his feedback at all.

Reply:
@d3lm Do you mean like this? Apologies if I'm not getting it. (it also seems like the fromPromise has been deprecated as I can't import it, a quick google search later and most people suggested i just use `from` )

  ```
checkUpdatesManually() {

    let appUpdate: Observable<any>;

    if (!appUpdate) {
      appUpdate = from(this.serviceWorkerUpdates.checkForUpdate());
      appUpdate.subscribe(() => {
        this.snackbar.open(`Check complete, You'll be notified if an update exists...`, 'OK', { duration: 2000 });
      });
    }
  }

```

Reply:
@d3lm, @KwintenP  Yes I had seen his suggestions, since I was going for the promise version, I couldn't apply them yet, but I've had another look since converting back to an Observable, i'm not sure what `shareReplay ` does to be honest, it seems tied to the interval which I've left out now, if it's still required I'll use it just let me know again.

I also forgot about converting the `map` pipe to `mapTo` but I've done that now.